File: ledger/ledger-common/src/main/java/net/corda/v5/ledger/common/transaction/TransactionVerificationException.java
package net.corda.v5.ledger.common.transaction;

import net.corda.v5.base.exceptions.CordaRuntimeException;
import net.corda.v5.crypto.SecureHash;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Indicates that some aspect of the transaction named by {@link #txId} violates some rules.
 */
public final class TransactionVerificationException extends CordaRuntimeException {
    private final SecureHash txId;

    public TransactionVerificationException(@NotNull SecureHash txId, @NotNull String message, @Nullable Throwable cause) {
        super(message + ", transaction: " + txId, cause);
        this.txId = txId;
    }

    /**
     * getTxId.
     * @return the Merkle root hash (identifier) of the transaction that failed verification.
     */
    @NotNull
    public SecureHash getTxId() {
        return txId;
    }
}

File: ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/transaction/TransactionVerificationException.kt
package net.corda.v5.ledger.common.transaction

import net.corda.v5.application.flows.exceptions.FlowException
import net.corda.v5.crypto.SecureHash

/**
 * Indicates that some aspect of the transaction named by [txId] violates the platform rules. The exact type of failure
 * is expressed using a subclass. TransactionVerificationException is a [FlowException] and thus when thrown inside
 * a flow, the details of the failure will be serialised, propagated to the peer and rethrown.
 *
 * @property txId the Merkle root hash (identifier) of the transaction that failed verification.
 */
abstract class TransactionVerificationException(val txId: SecureHash, message: String, cause: Throwable?)
    : FlowException("$message, transaction: $txId", cause)